Switch to our own Base64 implementation

davidebeatrici opened the pull request:
Written in modern C++, it strictly adheres to the Base64 specification.

This means that alternate alphabets and/or the lack of padding are rejected.

A benchmark would be nice to have, once we switch to a proper test framework.

Krzmbrzl commented:
> A benchmark would be nice to have, once we switch to a proper test framework.

-> https://github.com/google/benchmark
I have used that one before and it is pretty amazing.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,133 @@
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		constexpr std::array< uint8_t, 256 > table = {
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 62, 64, 64, 64, 63, 52, 53, 54, 55,
+			56, 57, 58, 59, 60, 61, 64, 64, 64, 64, 64, 64, 64, 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12,
+			13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 64, 64, 64, 64, 64, 64, 26, 27, 28, 29, 30, 31, 32,
+			33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+		};
```
What is that good for? Why have a table with 256 equal entries? :eyes:

davidebeatrici replied:
Not equal, look closely.

Krzmbrzl replied:
Oh indeed. Still - what is it good for?
Also we could probably use `std::array`'s `fill` function and then only set the fields that are != 64 manually afterwards.

davidebeatrici replied:
[`std::fill`](https://en.cppreference.com/w/cpp/container/array/fill) is only `constexpr` from C++20 onward though.

Please note that the values in the table can be calculated in case we want to get rid of it entirely.

Krzmbrzl replied:
I was referring to https://en.cppreference.com/w/cpp/container/array/fill, but for this the same constexpr restrictions apply.
But then again, we don't really need this function to be constexpr. I would say that the scenario in which you want to encode something as Base64 at compile time is somewhat rare, don't you think?

On the other hand, we could easily write our own constexpr `fill` function to make this code more readable.

davidebeatrici replied:
I replaced both tables with a function for each.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,133 @@
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
+		constexpr std::array< uint8_t, 256 > table = {
```
```suggestion
		constexpr std::array< uint8_t, std::numberic_limits< uint8_t >::max() > table = {
```

davidebeatrici replied:
Makes sense, as it's supposed to hold a value for every possible character.

davidebeatrici commented:
<details>
<summary>Windows build failure</summary>

```c++
Error: D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): error C2672: 'std::min': no matching overloaded function found
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(74): note: could be '_Ty std::min(std::initializer_list<_Elem>)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: '_Ty std::min(std::initializer_list<_Elem>)': expects 1 arguments - 2 provided
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(71): note: or       '_Ty std::min(std::initializer_list<_Elem>,_Pr)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: '_Ty std::min(std::initializer_list<_Elem>,_Pr)': could not deduce template argument for 'std::initializer_list<_Elem>' from 'unsigned long'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(64): note: or       'const _Ty &std::min(const _Ty &,const _Ty &) noexcept(<expr>)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: 'const _Ty &std::min(const _Ty &,const _Ty &) noexcept(<expr>)': template parameter '_Ty' is ambiguous
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: could be 'gsl::span<uint8_t,18446744073709551615>::size_type'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: or       'unsigned long'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: 'const _Ty &std::min(const _Ty &,const _Ty &) noexcept(<expr>)': could not deduce template argument for 'const _Ty &' from 'gsl::span<uint8_t,18446744073709551615>::size_type'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(54): note: or       'const _Ty &std::min(const _Ty &,const _Ty &,_Pr) noexcept(<expr>)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(73): note: 'const _Ty &std::min(const _Ty &,const _Ty &,_Pr) noexcept(<expr>)': expects 3 arguments - 2 provided
Error: D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): error C2672: 'std::min': no matching overloaded function found
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(74): note: could be '_Ty std::min(std::initializer_list<_Elem>)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: '_Ty std::min(std::initializer_list<_Elem>)': expects 1 arguments - 2 provided
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(71): note: or       '_Ty std::min(std::initializer_list<_Elem>,_Pr)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: '_Ty std::min(std::initializer_list<_Elem>,_Pr)': could not deduce template argument for 'std::initializer_list<_Elem>' from 'unsigned long'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(64): note: or       'const _Ty &std::min(const _Ty &,const _Ty &) noexcept(<expr>)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: 'const _Ty &std::min(const _Ty &,const _Ty &) noexcept(<expr>)': template parameter '_Ty' is ambiguous
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: could be 'gsl::span<const uint8_t,18446744073709551615>::size_type'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: or       'unsigned long'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: 'const _Ty &std::min(const _Ty &,const _Ty &) noexcept(<expr>)': could not deduce template argument for 'const _Ty &' from 'gsl::span<const uint8_t,18446744073709551615>::size_type'
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\utility(54): note: or       'const _Ty &std::min(const _Ty &,const _Ty &,_Pr) noexcept(<expr>)'
D:\a\libmumble\libmumble\include\mumble/Base64.hpp(122): note: 'const _Ty &std::min(const _Ty &,const _Ty &,_Pr) noexcept(<expr>)': expects 3 arguments - 2 provided
```
</details>

This is of course due to `size_t` being `unsigned long long` instead of `unsigned long`.

Unfortunately the proper literal was only introduced in C++23: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0330r8.html

We can get away with `size_t{ value }` though.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
 public:
-	class P;
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
```
I assume spans are also implicitly constructible from a `std::string`? Probably not from a `const char *` though. The latter is likely not particularly important though...

davidebeatrici replied:
They construct from both just fine (`StrViewConst` is used for `decode()`, `StrView` for `encode()`).

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
+class Base64 {
```
```suggestion
namespace base64 {
```
We should avoid using classes as namespaces.

davidebeatrici replied:
We have private members in `Base64`.

Krzmbrzl replied:
But we don't need to. If you want "private" definitions, use a nested `details` namespace. That's what is usually done.

davidebeatrici replied:
But why would we want to do that? I mean, what is the actual advantage?

Krzmbrzl replied:
Not using classes as namespaces. Classes are meant for defining instantiable objects, but here we don't want anything to be instantiable.

davidebeatrici replied:
It brings a few disadvantages though, such as members having to be declared beforehand and the need to specify that we want to use the `details` namespace in the function(s).

Krzmbrzl replied:
I would consider that to be less of a disadvantage than having an instantiable class flying around, that actually isn't meant to be instantiated :thinking:

davidebeatrici replied:
In my opinion it's actually a greater disadvantage, especially considering the compiler should optimize out the allocation of a class that's effectively 0 bytes.

Krzmbrzl replied:
This is about semantics, not performance. It literally makes zero sense to define a class that is not meant to be instantiated. What we are doing is 99% what namespaces are for.
Having private definitions isn't possible with namespaces, but what we are currently using it for is also unnecessary. I'm quite sure we can do without casting the span to an int-span, but instead keep using the byte-span.
And what would it hurt to have the pad character and the char to pos functions be public? And if you for some reason really think they should not be part of the public API, put them into a `details` nested namespace as everyone else seems to do as well.
If you are bothered by the fact that this requires defining the private stuff first, then simply outsource them into a separate header file in a `details` subdirectory and include that as the first thing. This will also completely remove those definitions from the eyes of everyone looking at the public API - which the class-approach can't do.

davidebeatrici replied:
> I'm quite sure we can do without casting the span to an int-span, but instead keep using the byte-span.

Already done.

> And what would it hurt to have the pad character and the char to pos functions be public? 

We can make them public now that the data types are consistent, no problem.

> If you are bothered by the fact that this requires defining the private stuff first, then simply outsource them into a separate header file in a `details` subdirectory and include that as the first thing. This will also completely remove those definitions from the eyes of everyone looking at the public API - which the class-approach can't do.

And what if we want to define public stuff in a specific order?

Using a namespace would force us to move `PAD_CHAR`, `charToByte()` and `byteToChar()` at the very top unless we make use of forward declarations.

Also, we may eventually find a case with two (or more) functions depending on each other.

Krzmbrzl replied:
> And what if we want to define public stuff in a specific order?

Why would you want to do that? And if you really want to, you can always declare functions first - in whatever order you want - and then  define them later (or in a separate file).

 > Also, we may eventually find a case with two (or more) functions depending on each other.

No problem: first define one and then the other. And if A depends on B and B depends on A, you've most likely introduced shady, if not buggy code anyway.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+	using StrView      = gsl::span< char >;
+	using StrViewConst = gsl::span< const char >;
+
+	static constexpr size_t decode(const BufView out, StrViewConst in) {
```
Making an output parameter `const` seems wrong.

Furthermore, we should add an overload that will return e.g. a `std::vector` containing the decoded data. It can internally simply call this function, but API-wise that'll be more straight forward as long as you don't need a specific container for the out data.

davidebeatrici replied:
> Making an output parameter `const` seems wrong.

It's the view itself to be `const`, not the actual buffer (that would be `BufViewConst`).

> Furthermore, we should add an overload that will return e.g. a `std::vector` containing the decoded data. It can internally simply call this function, but API-wise that'll be more straight forward as long as you don't need a specific container for the out data.

Yes, we will definitely add overloads/helpers in the future.

Krzmbrzl replied:
> It's the view itself to be const, not the actual buffer (that would be BufViewConst).

I know. That's why I said "seems wrong" and not "is a bug" :)
It is still semantically weird.

davidebeatrici replied:
Yeah, I can see why.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+		};
+
+		if (in.empty() || in.size() % 4) {
+			return 0;
+		}
```
In case the size is not a multiple of 4, we should throw an exception as this is clearly an error that shouldn't simply be ignored.

davidebeatrici replied:
The returned size should always be checked though. I agree that an exception would be helpful in knowing the cause behind the failure.

Krzmbrzl replied:
Yeah, absolutely. Size-checking is essential. Just what is done upon an invalid size is what I would like to see changed :)

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			return 0;
+		}
+
+		size_t size = in.size() / 4 * 3;
```
```suggestion
		size_t size = (in.size() / 4) * 3;
```
This makes precedence clearer

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+		}
+
+		size_t size = in.size() / 4 * 3;
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2 : 1;
+		}
```
This should probably have a comment explaining what is going on. Also, couldn't the input have up to three padding characters rather than only one or two?

davidebeatrici replied:
Nope, as they're only used to make the string a multiple of 4.

Krzmbrzl replied:
Ah right - if I would add a padding of three, that means my original value is already a multiple of three :+1:

davidebeatrici replied:
Yep!

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			size -= *++iter == PAD_CHAR ? 2 : 1;
+		}
+
+		if (!out.data()) {
+			return size;
+		}
```
This should be an error (exception).
We should provide a separate, dedicated function to obtain the size of to-be decoded data (e.g. `decodedSize(...)`)

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			return size;
+		}
+
+		if (out.size() < size) {
+			return 0;
+		}
```
This seems like odd behavior. Either this is an error and we should throw an exception, or the function should decoded as much as fits into the output buffer and then return how many characters of the input stream this decoding spanned.

davidebeatrici replied:
Partial output would be nice to have, I would do it in a separate PR though.

Krzmbrzl replied:
Alright. Then we should probably also add a separate function for it so one can be sure to get an exception if the provided buffer was not large enough when using one function and if one wants to make use of partial encoding, one can use the other function. Internally, they can share the same implementation.

davidebeatrici replied:
Yes, absolutely.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			return 0;
+		}
+
+		ByteView data(reinterpret_cast< uint8_t * >(out.data()), out.size());
```
Why the change of type? `std::byte` seems like the more appropriate type anyway.

davidebeatrici replied:
Operations on `std::byte` are not allowed.

Krzmbrzl replied:
What operations?
To be honest, the more I use `std::byte`, the less I like it. Maybe we would be better off by simply using `byte` as an alias for `uint8_t`?

davidebeatrici replied:
Arithmetic operations.

`std::byte` is meant to represent data that is in an unknown/unspecified format, which is actually the case here.

However, as we know what we need to do with the data, we cast to the type that is more comfy for us to work with.

Krzmbrzl replied:
Bit operations in bytes are quite common though, so it is silly that this is not allowed 👀

davidebeatrici replied:
Well, that's why [`std::to_integer()`](https://en.cppreference.com/w/cpp/types/byte#std::to_integer) exists.

The main reason would be to avoid dangerous implicit casts.

Krzmbrzl replied:
Ah, but according to [that site](https://en.cppreference.com/w/cpp/types/byte) the bitwise operators are overloaded for `std::byte`, so we shouldn't have issues in the first place? Can't we replace the addition by a binary AND?

davidebeatrici replied:
Yup.

davidebeatrici replied:
Better idea:

```c++
static constexpr std::byte operator+(const std::byte lhs, const std::byte rhs) {
	const auto ret = std::to_integer< uint8_t >(lhs) + std::to_integer< uint8_t >(rhs);
	return std::byte(ret);
}
```

Krzmbrzl replied:
That'd also solve the problem, but imo it'd be better to use a binary AND, if that is what is being performed. A binary AND and an addition are semantically very different things, so I think one should make it explicit what is going on :thinking:

davidebeatrici replied:
It's really an addition, a binary AND would give you a different result.

Krzmbrzl replied:
Ho - indeed. Yeah then using the overload for `operator+` seems good to me :+1:

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			data[0] = (table[in[0]] << 2) + ((table[in[1]] & 0x30) >> 4);
+
+			if (in[2] == '=') {
+				break;
+			}
+
+			data[1] = ((table[in[1]] & 0x0f) << 4) + ((table[in[2]] & 0x3c) >> 2);
+
+			if (in[3] == '=') {
+				break;
+			}
 
-	Base64();
-	virtual ~Base64();
+			data[2] = ((table[in[2]] & 0x03) << 6) + table[in[3]];
 
-	virtual explicit operator bool();
+			in   = in.subspan(4);
+			data = data.subspan(std::min(size_t{ 3 }, data.size()));
```
This requires a comment what is going on

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			data[2] = ((table[in[2]] & 0x03) << 6) + table[in[3]];
 
-	virtual explicit operator bool();
+			in   = in.subspan(4);
+			data = data.subspan(std::min(size_t{ 3 }, data.size()));
```
What is the advantage of using this sub-spabn approach to a more traditional for-loop? Technically, this subspan view requires constant creation of temporary objects and also a for-loop makes it more obvious that this is not an indeterminant loop, but one that has a precise iteration count.

davidebeatrici replied:
In this specific case the sub-span is useful because it scales the size automatically while advancing the pointer.

Krzmbrzl replied:
On a quick glance we don't seem to access the size inside the loop, do we? :eyes:

davidebeatrici replied:
Sorry, we do it in `encode()`. In `decode()`'s loop we could indeed avoid reassigning the span and simply advancing the pointer.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,134 @@
+			return 0;
+		}
+
+		const size_t size = static_cast< size_t >(std::ceil(static_cast< double >(in.size()) / 3) * 4);
```
`std::ceil` being `constexpr` seems is non-standard for C++ < v23, so we should not use this here.
```suggestion
		// (a + 2) / 3 == std::ceil(static_cast< double >(a) / 3), but constexpr
		const size_t size = 4 * ((in.size() + 2) / 3);
```

Krzmbrzl reviewed (changes requested):
Essentially all comments on the `decode` function also apply to the `encode` one.

Krzmbrzl commented:
Oh and another thing to consider: As you want to implement this header-only anyway, we could also template the functions to allow any type of the correct size (or anything that has at least the appropriate size). That way, we wouldn't have the `uint8_t` vs `std::byte` vs `char` vs ... problem

davidebeatrici commented:
The failing Windows builds are probably the due to a missing `<stdexcept>` include.

> Oh and another thing to consider: As you want to implement this header-only anyway, we could also template the functions to allow any type of the correct size (or anything that has at least the appropriate size). That way, we wouldn't have the `uint8_t` vs `std::byte` vs `char` vs ... problem

Or rather, template wrappers for them: thanks to the spans we can get away with simple casts.

Krzmbrzl commented:
> thanks to the spans we can get away with simple casts.

Doesn't seem like it: https://godbolt.org/z/5P5dWc5nj :eyes:

davidebeatrici commented:
Sorry, by "simple" I meant casting the element type and specifying the size accordingly.

Krzmbrzl commented:
Maybe we should define our own `span_cast` that does this under the hood. That will probably make our code a bit cleaner

davidebeatrici commented:
Indeed, I was actually thinking about it. There are quite a few places in our codebase where it would be useful.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+			throw std::runtime_error("Input is not a multiple of 4!");
+		}
+
+		const auto size = decodedSize(in);
```
Intransparent use of `auto`

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+			throw std::runtime_error("Insufficient output buffer size!");
+		}
+
+		for (auto data = reinterpret_cast< uint8_t * >(out.data()); !in.empty(); data += 3, in = in.subspan(4)) {
```
Why not simply use a regular for-loop that does
```cpp
for (std::size_t i = 0; i < in.size(); ++i)
```
To me this would still seem like the most readable solution :thinking: 
All offsets/sizes can then be calculated based on the value of `i`

davidebeatrici replied:
It's prone to errors, also consider that we are advancing two separate iterators by a specific amount of bytes and then checking the remaining size in the code.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+		// Base64 can have up to 2 pad characters, in order to make the string a multiple of 4.
+		// In order to calculate the exact output size we have to ignore padding.
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2U : 1U;
```
```suggestion
			size -= *++iter == PAD_CHAR ? 2 : 1;
```
why using extra suffixes?

davidebeatrici replied:
Throws a signed/unsigned warning otherwise.

Krzmbrzl replied:
Ah indeed

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+		return size;
+	}
+
+	static constexpr size_t encodedSize(const BufViewConst in) { return ((in.size() + 2) / 3) * 4; }
```
```suggestion
	static constexpr size_t encodedSize(const BufViewConst in) {
		// ceil((float) a / 3) == (a + 2) / 3, assuming a is an integer
		return ((in.size() + 2) / 3) * 4;
	}
```

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+		if (ch == '/' || ch == '_')
+			return 63;
+
+		throw std::range_error("");
```
We might want to consider using [Hedley](https://github.com/nemequ/hedley) and replacing that throw-statement with `HEDLEY_UNREACHABLE` (prefixed by an `assert(false)` to trap in debug builds)

davidebeatrici replied:
The line is reachable in case of invalid Base64 characters though.

Krzmbrzl replied:
That's what the assert is for. And the validity of the input should be checked before calling the function anyway.

But if you want the error handling to happen here, then the exception will require a meaningful message (that includes the offending character)

davidebeatrici replied:
> And the validity of the input should be checked before calling the function anyway.

We could eventually add a function performing the check, but its use shouldn't be forced on the API user.

> But if you want the error handling to happen here, then the exception will require a meaningful message (that includes the offending character)

Right.

Krzmbrzl replied:
Yeah :+1:

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+			return static_cast< uint8_t >(ch) - 'a' + ('Z' - 'A') + 1;
+		if (ch >= '0' && ch <= '9')
+			return static_cast< uint8_t >(ch) - '0' + ('Z' - 'A') + ('z' - 'a') + 2;
+		if (ch == '+' || ch == '-')
+			return 62;
+		if (ch == '/' || ch == '_')
+			return 63;
```
Are `+`, `-`  and `/`, `_` really equivalent in the encoding? :eyes:

davidebeatrici replied:
Yes, different alphabet (standard vs URL-safe).

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+
+		const auto size = decodedSize(in);
+		if (out.size() < size) {
+			throw std::runtime_error("Insufficient output buffer size!");
```
Should we perhaps throw a [std::invalid_argument](https://cplusplus.com/reference/stdexcept/invalid_argument/) exception?

davidebeatrici replied:
Good idea.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+	using ByteView      = gsl::span< uint8_t >;
+	using ByteViewConst = gsl::span< const uint8_t >;
+
+	static constexpr uint8_t charToPos(const char ch) {
```
```suggestion
	static constexpr uint8_t charToNumeric(const char ch) {
```
Is this a better name? `pos` seems to indicate that we were using the result to index something, which is not the case :thinking:

davidebeatrici replied:
Right, it's not a position in a table.

davidebeatrici replied:
Is `charToByte()` alright?

Krzmbrzl replied:
That would be a perfect fit, but I think then the function should return a `std::byte` instead of a `uint8_t` :thinking: 
Afaict this shouldn't cause issues on the calling site :eyes:

davidebeatrici replied:
Yeah, I did that.

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,30 +3,147 @@
+		throw std::range_error("");
+	}
+
+	static constexpr char posToChar(const uint8_t pos) {
```
```suggestion
	static constexpr char numericToChar(const uint8_t num) {
		assert(num <= 63);
```

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
+		if (num == 63)
+			return '/';
+
+		throw std::range_error(std::string() + num + " is not a valid byte!");
```
```suggestion
		throw std::range_error(std::to_string(+num) + " is not a valid byte that can be translated into Base64 alphabet!");
```

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
+			return 0;
+		}
+
+		if (in.size() % 4) {
```
```suggestion
		if (in.size() % 4 != 0) {
```
makes it more obvious to read

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
+		}
+
+		for (; !in.empty(); out = out.subspan(3), in = in.subspan(4)) {
+			out[0] = (charToByte(in[0]) << std::byte(2)) + ((charToByte(in[1]) & std::byte(0x30)) >> 4);
```
```suggestion
			out[0] = (charToByte(in[0]) << 2) + ((charToByte(in[1]) & std::byte(0x30)) >> 4);
```

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
+			}
+
+			out[1] =
+				((charToByte(in[1]) & std::byte(0x0f)) << std::byte(4)) + ((charToByte(in[2]) & std::byte(0x3c)) >> 2);
```
```suggestion
				((charToByte(in[1]) & std::byte(0x0f)) << 4) + ((charToByte(in[2]) & std::byte(0x3c)) >> 2);
```

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
+				break;
+			}
+
+			out[2] = ((charToByte(in[2]) & std::byte(0x03)) << std::byte(6)) + charToByte(in[3]);
```
```suggestion
			out[2] = ((charToByte(in[2]) & std::byte(0x03)) << 6) + charToByte(in[3]);
```

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
 
-	Base64();
-	virtual ~Base64();
+			if (in.size() >= 2) {
+				out[1] = byteToChar(((in[0] & std::byte(0x03)) << 4) + ((in[1] & std::byte(0xf0)) >> 4));
 
-	virtual explicit operator bool();
+				if (in.size() >= 3) {
+					out[2] = byteToChar(((in[1] & std::byte(0x0f)) << 2) + ((in[2] & std::byte(0xc0)) >> 6));
+					out[3] = byteToChar(in[2] & std::byte(0x3f));
+				} else {
+					out[2] = byteToChar((in[1] & std::byte(0x0f)) << 2);
+					out[3] = PAD_CHAR;
+				}
+			} else {
+				out[1] = byteToChar((in[0] & std::byte(0x03)) << 4);
+				out[2] = PAD_CHAR;
+				out[3] = PAD_CHAR;
+			}
 
-	virtual size_t decode(const BufView out, const BufViewConst in);
-	static size_t encode(const BufView out, const BufViewConst in);
+			// All data blocks are guaranteed to be 3 bytes except the last one.
+			in = in.subspan(std::min(size_t{ 3 }, in.size()));
+		}
```
Analogous to the `decode` function, a comment textually explaining how this stuff works would be good.

Krzmbrzl reviewed (commented):
Only some small stuff is left. Once this has been addressed, I think this implementation is ready to go :+1:

Krzmbrzl commented on include/mumble/Base64.hpp:
```diff
@@ -3,31 +3,175 @@
+		return std::byte(ret);
+	}
+
+	// The standard doesn't overload this operator with both operands as std::byte...
+	static constexpr std::byte operator<<(const std::byte lhs, const std::byte rhs) {
+		return lhs << std::to_integer< uint8_t >(rhs);
+	}
+
```
```suggestion
```
This is not needed if you apply my suggestions from above

davidebeatrici replied:
It's required, otherwise we end up with compile warnings due to `int` being the implicit data type.

Krzmbrzl reviewed (commented):
Oh and of course: All functions require documentation before merging

I would suggest JavaDoc-style Doxygen comments as those are much easier to distinguish from regular comments than blocks of `///` comments. E.g. like this
```
/**
 * Description of function goes here
 * 
 * @param in The buffer to decode
 */
```

davidebeatrici committed:
Switch to our own Base64 implementation

Written in modern C++, it strictly adheres to the Base64 specification.

This means that the lack of padding is rejected when decoding.
The alternate alphabet that is used for URLs is handled just fine.

A benchmark would be nice to have, once we switch to a proper test framework.

Krzmbrzl commented:
Just to be explicit: The only thing missing from my POV is the documentation (+ the mentioned in-source explanatory comments)
